### PR TITLE
Dummy Contract Env And Interpreter

### PR DESCRIPTION
**Pull request by mariari**
Gets the Interpreter for Michelson up. To a state where we can easily use it.

### CODE REVIEW

**Comment by mariari** [comment]
> Semantic test-cases for compiled contracts (using the interpreter)

this part is now in and being done!

File: src/Juvix/Backends/Michelson/Compilation/Types.hs [hunk 1]
```diff
@@ -52,10 +53,21 @@ data CompilationLog
   | OptimisedByMorley SomeInstr SomeInstr
   deriving (Generic, Show)
 
+data EmptyInstr where
```
Comment (cwgoes) [inline src/Juvix/Backends/Michelson/Compilation/Types.hs:56]: Should we clean these up?

File: src/Juvix/Backends/Michelson/Parameterisation.hs [hunk 2]
```diff
@@ -37,6 +44,52 @@ constType v =
...
+              let newTerm =
+                    Run.applyPrimOnArgs (Prim.toAnn fun) (Prim.toAnn <$> args)
+                  -- TODO ∷ do something with the logs!?
+                  (compd, _log) = Compilation.compileExpr newTerm
+               in case compd >>= Interpreter.dummyInterpret of
+                    Right x ->
+                      Constant x
+                        |> Prim.Return
+                        |> Right
+                    -- TODO :: promote this error
```
Comment (cwgoes) [inline src/Juvix/Backends/Michelson/Parameterisation.hs:84]: I believe that this failure indicates a mistake in the typechecker, which should have been run already by this stage

File: test/Backends/Michelson.hs [hunk 3]
```diff
@@ -19,24 +20,44 @@ import Prelude (show)
...
+runExpr term =
+  fst (compileExpr term)
+
+runContractWrap :: Term -> Type -> Either DSL.CompError (Contract' ExpandedOp)
+runContractWrap term ty =
+  runContract (Ann zero newTy (J.LamM [] ["gen%%%"] term)) newTy
+  where
+    newTy = J.Pi zero (primTy unitPair) ty
+
+shouldCompExp :: AnnTerm PrimTy NewPrim -> M.Value -> T.TestTree
```
Comment (cwgoes) [inline test/Backends/Michelson.hs:37]: Confusing name, but OK

File: src/Juvix/Backends/Michelson/Parameterisation.hs [hunk 4]
```diff
@@ -37,6 +44,52 @@ constType v =
...
+    (Prim.Return PrimTy PrimVal)
+applyProper fun args =
+  case Prim.term fun of
+    Constant _i ->
+      case length args of
+        0 ->
+          Right (Prim.Return (Prim.term fun))
+        _x ->
+          Left (CoreTypes.TypecheckerError "Applied a constant to argument")
+    Inst instruction ->
```
Comment (cwgoes) [inline src/Juvix/Backends/Michelson/Parameterisation.hs:61]: Somewhere we need to deal with dependency, namely, polymorphic functions in Michelson will be dependent in Juvix, e.g. `ADD`:

```idris
add : (t : Type) -> t -> t -> t
```

— so we need to deal with the first application to `t` (e.g. `Int` or `Tez`) somewhere, how do you think we should do that?

File: package.yaml [hunk 5]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
 dependencies:
   - base >= 4.11 && < 5
   - protolude
+  - directory
+  - lens
+  - Cabal
+  - extensible-data
+  - capability >= 0.2.0.0
+  ###########################
+  # Depedency for OS
```
Comment (cwgoes) [inline package.yaml:25]:
```suggestion
  # Dependency for OS
```

File: package.yaml [hunk 6]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
...
+  ###########################
+  - turtle
+  - time
+  ###########################
+  # Depedency for Arithmetic
```
Comment (cwgoes) [inline package.yaml:30]:
```suggestion
  # Dependency for Arithmetic
```

File: package.yaml [hunk 7]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
...
+  ###########################
+  - bulletproofs
+  - elliptic-curve
+  - sonic
+  ###########################
+  # Depedency for Structs
```
Comment (cwgoes) [inline package.yaml:36]:
```suggestion
  # Dependency for Structs
```

File: package.yaml [hunk 8]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
...
   - containers
   - unordered-containers
   - hashable
   - fgl
-  - lens
+  - fgl-visualize
+  - bytestring
+  - enummapset
+  ###########################
+  # Depedency for Parsers
```
Comment (cwgoes) [inline package.yaml:48]:
```suggestion
  # Dependency for Parsers
```

File: package.yaml [hunk 9]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
...
   - attoparsec
   - attoparsec-expr
-  - capability >= 0.2.0.0
-  - fgl-visualize
-  - enummapset
-  - turtle
-  - directory
+  - aeson
+  ###########################
+  # Depedency for LLVM
```
Comment (cwgoes) [inline package.yaml:55]:
```suggestion
  # Dependency for LLVM
```

File: package.yaml [hunk 10]
```diff
@@ -16,33 +16,54 @@ extra-source-files:
...
+  ###########################
   - llvm-hs
   - llvm-hs-pure
   - llvm-hs-pretty
-  - bytestring
-  - time
+  ###########################
+  # Depedency for Morley
```
Comment (cwgoes) [inline package.yaml:61]:
```suggestion
  # Dependency for Morley
```

File: src/Juvix/Backends/Michelson/Parameterisation.hs (hunk 2 above)
Comment (mariari) [reply src/Juvix/Backends/Michelson/Parameterisation.hs:84]: I'd still like to reserve the errors of the compiler itself, and so an Either over Maybe is nicer?

File: src/Juvix/Backends/Michelson/Compilation/Types.hs (hunk 1 above)
Comment (mariari) [reply src/Juvix/Backends/Michelson/Compilation/Types.hs:56]: Yes I'll remove them

**Comment by cwgoes** [comment]
Per discussion, we need to add monomorphic version of all primitives, e.g. `add_int_int`, `add_nat_nat`.

File: test/Backends/Michelson.hs (hunk 3 above)
Comment (mariari) [reply test/Backends/Michelson.hs]: Changed to interpretExpression

File: src/Juvix/Backends/Michelson/Parameterisation.hs (hunk 2 above)
Comment (mariari) [reply src/Juvix/Backends/Michelson/Parameterisation.hs:85]: This is now promoted, however I changed the 

Juvix.Core.Types.TermAssignment

**Pull request merged**